File: 37-sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        rows = [set() for _ in range(9)]  
        cols = [set() for _ in range(9)]  
        boxes = [set() for _ in range(9)]  

        
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    num = board[i][j]
                    rows[i].add(num)
                    cols[j].add(num)
                    boxes[(i // 3) * 3 + (j // 3)].add(num)

        def backtrack(row, col):
            if row == 9: 
                return True
            if col == 9:  
                return backtrack(row + 1, 0)
            if board[row][col] != ".":  
                return backtrack(row, col + 1)

            box_index = (row // 3) * 3 + (col // 3)  
            for num in "123456789":
                if num not in rows[row] and num not in cols[col] and num not in boxes[box_index]:
                    
                    board[row][col] = num
                    rows[row].add(num)
                    cols[col].add(num)
                    boxes[box_index].add(num)

                    if backtrack(row, col + 1):  
                        return True

                    
                    board[row][col] = "."
                    rows[row].remove(num)
                    cols[col].remove(num)
                    boxes[box_index].remove(num)

            return False

        backtrack(0, 0)

File: 37-sudoku-solver/sudoku-solver.java
class Solution {
    private Set<Character>[] row = new HashSet[9];
    private Set<Character>[] col = new HashSet[9];
    private Set<Character>[] box = new HashSet[9];
    private char[][] grid;

    private boolean backtrack(int i, int j) {
        if (i == 9) {
            return true;
        }
        if (j == 9) {
            return backtrack(i + 1, 0);
        }
        if (grid[i][j] != '.') {
            return backtrack(i, j + 1);
        }

        int box_index = (i / 3) * 3 + (j / 3);
        for (char c = '1'; c <= '9'; c++) {
            if (!row[i].contains(c) && !col[j].contains(c) && !box[box_index].contains(c)) {
                grid[i][j] = c;
                row[i].add(c);
                col[j].add(c);
                box[box_index].add(c);

                if (backtrack(i, j + 1)) {
                    return true;
                }

                grid[i][j] = '.';
                row[i].remove(c);
                col[j].remove(c);
                box[box_index].remove(c);
            }
        }
        return false;
    }

    public void solveSudoku(char[][] board) {
        grid = board;
        for (int i = 0; i < 9; i++) {
            row[i] = new HashSet<>();
            col[i] = new HashSet<>();
            box[i] = new HashSet<>();
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                char num = board[i][j];
                row[i].add(num);
                col[j].add(num);
                box[(i / 3) * 3 + (j / 3)].add(num);
            }
        }
        backtrack(0, 0);
    }
}